### PR TITLE
Bug #18469: fix Link::serialize method

### PR DESCRIPTION
**Pull request by ggh2e3**
| Q             | A
| ------------- | ---
| Is bugfix?  ✔️
| New feature? ❌
| Breaks BC? ❌
| Fixed issues  | #18469 

Hi, I applied the fix proposed above with some unit tests. Hope it helps

### CODE REVIEW

File: framework/web/Link.php [hunk 1]
```diff
@@ -66,7 +66,11 @@ public static function serialize(array $links)
                     $link[$i] = $l instanceof self ? array_filter((array) $l) : ['href' => $l];
                 }
                 $links[$rel] = $link;
-            } elseif (!$link instanceof self) {
+            }
+            if ($link instanceof self) {
```
[the remark this answers is not in the document]
Comment (ggh2e3) [reply framework/web/Link.php:70]: something like this?
`        foreach ($links as $rel => $link) {
            if (is_array($link)) {
                $link = self::serialize($link); // refactored snippet
                $links[$rel] = $link;
            }
            if ($link instanceof self) {
                $l = array_filter((array)$link);
                $links[$rel] = $l;
            } else {
                $links[$rel] = ['href' => $link];
            }
        }`

[the remark this answers is not in the document]
Comment (ggh2e3) [reply framework/web/Link.php:70]: Are you referring to line 66(or maybe the whole method)?
I'd replace line 66 with this recursive call` $link = self::serialize($link);` in order to remove the second loop

**Comment by rob006** [comment]
It is only me, or these test do not have sense at all? :P

**Comment by bizley** [comment]
Heh, yes, this is what I thought as well, so I had to go through the code again.
@ggh2e3 could you refactor the tests to not confuse? For example [this](https://github.com/yiisoft/yii2/pull/20020/files#diff-649b7c9c42053caeda40e96fad0d665907b4b281093f719df437f138c532698bR29) looks like a config for a single object so the expected result is weird.

**Comment by ggh2e3** [comment]
Yep, I can refactor the tests to avoid confusion.

**Comment by ggh2e3** [comment]
@bizley is it fine now?

File: framework/web/Link.php [hunk 2]
```diff
@@ -62,11 +62,13 @@ public static function serialize(array $links)
     {
         foreach ($links as $rel => $link) {
             if (is_array($link)) {
-                foreach ($link as $i => $l) {
-                    $link[$i] = $l instanceof self ? array_filter((array) $l) : ['href' => $l];
-                }
+                $link = self::serialize($link);
                 $links[$rel] = $link;
```
Comment (bizley) [inline framework/web/Link.php:66]:
```suggestion
                $links[$rel] = self::serialize($link);
```

File: framework/web/Link.php [hunk 3]
```diff
@@ -62,11 +62,13 @@ public static function serialize(array $links)
...
-            } elseif (!$link instanceof self) {
+            }
+            else if ($link instanceof self) {
+                $l = array_filter((array)$link);
+                $links[$rel] = $l;
```
Comment (bizley) [inline framework/web/Link.php:70]:
```suggestion
            } elseif ($link instanceof self) {
                $links[$rel] = array_filter((array)$link);
```

**Review by bizley — CHANGES_REQUESTED** [review]

**Review by bizley — APPROVED** [review]

**Review by Arhell — APPROVED** [review]

**Comment by samdark** [comment]
👍

**Pull request merged**